ranger-discovery-bundle/src/main/java/io/appform/ranger/discovery/bundle/id/DistributedIdGenerator.java

```diff
@@ -0,0 +1,288 @@
+     * @param prefix     String prefix
+     * @param domain     Domain for constraint selection
+     * @param skipGlobal Skip global constrains and use only passed ones
+     * @return Id if it could be generated
+     */
+    public Optional<Id> generateWithConstraints(final String prefix, final String domain, final boolean skipGlobal) {
+        val targetPartitionId = getTargetPartitionId(DOMAIN_SPECIFIC_CONSTRAINTS.getOrDefault(domain, Collections.emptyList()), skipGlobal);
+        return targetPartitionId.map(id -> generateForPartition(prefix, id));
+    }
+
+    public Optional<Id> generateWithConstraints(final String prefix,
+                                                final List<KeyValidationConstraint> inConstraints,
+                                                final boolean skipGlobal) {
+        val targetPartitionId = getTargetPartitionId(inConstraints, skipGlobal);
+        return targetPartitionId.map(id -> generateForPartition(prefix, id));
+    }
+
+    /**
+     * Generate id by parsing given string
+     *
+     * @param idString String idString
+     * @return Id if it could be generated
+     */
+    public Optional<Id> parse(final String idString) {
+        if (idString == null
+                || idString.length() < MINIMUM_ID_LENGTH) {
+            return Optional.empty();
+        }
+        try {
+            val matcher = PATTERN.matcher(idString);
+            if (matcher.find()) {
+                return Optional.of(Id.builder()
+                        .id(idString)
+                        .node(Integer.parseInt(matcher.group(3)))
+                        .exponent(Integer.parseInt(matcher.group(4)))
+                        .generatedDate(DATE_TIME_FORMATTER.parseDateTime(matcher.group(2)).toDate())
+                        .build());
+            }
+            return Optional.empty();
+        }
+        catch (Exception e) {
+            log.warn("Could not parse idString {}", e.getMessage());
+            return Optional.empty();
+        }
+    }
+
+    private int getTargetPartitionId() {
+        return SECURE_RANDOM.nextInt(partitionCount);
+    }
+
+    private Optional<Integer> getTargetPartitionId(final List<KeyValidationConstraint> inConstraints, final boolean skipGlobal) {
+//      ToDo: Check if we need Collision Checker here
+        return Optional.ofNullable(
+                RETRIER.get(() -> SECURE_RANDOM.nextInt(partitionCount)))
+                .filter(key -> validateId(inConstraints, key, skipGlobal));
+    }
+
+    protected boolean validateId(final List<KeyValidationConstraint> inConstraints,
+                                 final int partitionId,
+                                 final boolean skipGlobal) {
+        //First evaluate global constraints
+        val failedGlobalConstraint
+                = skipGlobal
+                  ? null
+                  : GLOBAL_CONSTRAINTS.stream()
+                          .filter(constraint -> !constraint.isValid(partitionId))
```

Comment:
Here we are assuming that constraints will always be on partitionId, Should we have it on ID level also, else client have to implement this method, and can't moveover seamlessly. Let's discuss it once.